### DISCUSSION TITLE
Embedding Unity Build into a UI Framework (Uno Platform)

### DISCUSSION THREAD

**Discussion by BertrandDeSaintBezier** [microsoft/microsoft-ui-xaml#8498] [Q&A]
Still new to C#/XAML developement. Unity (the game engine) provides a way of [embedding a build into a Windows application via Window Handles](https://docs.unity3d.com/Manual/UnityasaLibrary-Windows.html) (HWND), but I'm still trying to wrap my head around the concept of Handles as a whole. I'm currently working in Uno Platform and child components don't seem to have Handles, only the whole Window has one.

The code below embeds the Unity build into the entire window, but I would like it to only occupy a portion of it, such as containing it in a grid or canvas/panel - a "child window" of sorts. I've seen a lot of posts suggesting that I use SwapGridPanels instead to render the content.

Any ideas or pointers (haha) 🤔 ?

```csharp
public MainPage()
{
    this.InitializeComponent();

    try
    {
        var process = new Process();
        process.StartInfo.WorkingDirectory = "path to working dir...";
        process.StartInfo.FileName = "UnityApp.exe";

        var hWnd = WinRT.Interop.WindowNative.GetWindowHandle(App._window);
        process.StartInfo.Arguments = $"-parentHWND {hWnd.ToInt32()} {Environment.CommandLine}";
        process.StartInfo.UseShellExecute = true;
        process.StartInfo.CreateNoWindow = true;

        process.Start();
        process.WaitForInputIdle();
    }
    catch (Exception ex)
    {
        Console.Write(ex.Message);
    }
}
```

**Comment by DarranRowe** [comment]
The way Xaml works is that it uses its own implementation of the [Windows Visual Layer](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/visual-layer-in-desktop-apps) to do the rendering. The Visual Layer is implemented using [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/why-use-directcomposition-).
So, what does this mean? Well, only the top level window has the handle. Anything else is a visual in a tree, and a visual is just a bitmap stored in memory that is composed into what you finally see when the entire desktop is put together. So yes, only the top level window has a handle and there will be issues if you use this to target the window.
If you are able to set unity to target a swap chain instead of a HWND, then this is how you want to render content to the window. The [SwapChainPanel](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.swapchainpanel?view=windows-app-sdk-1.3) allows you to put a DXGI swap chain in the tree, but this does require more C++ interop. I'm sure there are potentially weird ways of doing this and targeting a HWND, but the swap chain panel is the best method which WinUI 3 supports.

**Reply by BertrandDeSaintBezier** [reply]
Solid answer, thanks for taking the time to explain this clearly :D I'm still exploring a couple of way to go about this. [Unity allows you to build your project as a UWP/XAML project](https://stackoverflow.com/questions/71673655/embedding-unity-in-wpf-winui-uwp-application-with-control-overlay) and has some custom methods to initialize a SwapChain, but we're looking into doing this using a .exe file and WinUI 3, (aka not not a UWP-style project).

I'll be sure to post updates as I secure a solution.
